presto-common/src/main/java/com/facebook/presto/common/type/AbstractVariableSignatureType.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common.type;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.block.UncheckedBlock;
+import io.airlift.slice.Slice;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * An abstract type that doesn't cache the type signature, because it may change dynamically
+ * due to distinct type lazy loading. However, all such type signatures will be equivalent and will work with
+ * equals() and give consistent hashes.
+ */
+public abstract class AbstractVariableSignatureType
+        implements Type
```

Comment:
Is it possible to have the class
- renamed to `AbstractComplexType` and
- inheriting from `AbstractType`?

Reply:
AbstractType caches type signature and marks getTypeSignature() as final. So we cannot inherit from AbstractType.

Do you want to call this one AbstractBasicType(i know, but it stores fewer things than AbstractType) and have AbstractType inherit from this one(as that one does more things)

Reply:
How about:
- `AbstractType` to have no type signature and 
- `AbstractPrimitiveType` to inherit `AbstractType` to have type signature

Then we have complex type to inherit `AbstractType` and all other types to inherit `AbstractPrimitiveType`